File: client/src/app/modules/quotation/services/service.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { AppRestService } from 'src/app/services/appRest.service';
import { Service } from '../../quotation/model/Service';
import { Affaire } from '../model/Affaire';
import { ServiceType } from '../model/ServiceType';

@Injectable({
  providedIn: 'root'
})
export class ServiceService extends AppRestService<Service> {

  constructor(http: HttpClient) {
    super(http, "quotation");
  }

  getServices() {
    return this.getList(new HttpParams(), "services");
  }

  addOrUpdateService(service: Service) {
    return this.addOrUpdate(new HttpParams(), "service", service, "Enregistré", "Erreur lors de l'enregistrement");
  }

  deleteService(service: Service) {
    return this.get(new HttpParams().set("serviceId", service.id), "service/delete", "Service supprimé", "Erreur lors de la suppression");
  }

  getServiceForServiceTypeAndAffaire(serviceType: ServiceType, affaire: Affaire) {
    return this.get(new HttpParams().set("serviceTypeId", serviceType.id).set("idAffaire", affaire.id), "service-type/provision");
  }

  getServiceForMultiServiceTypesAndAffaire(serviceTypes: ServiceType[], customLabel: string | undefined, affaire: Affaire) {
    const ids = serviceTypes.map(st => st.id).join(',');
    return this.getList(new HttpParams().set("customLabel", customLabel + "").set("serviceTypeIds", ids).set("idAffaire", affaire.id + ""), "service-types/provisions");
  }

  modifyServiceType(service: Service, serviceTypes: ServiceType[]) {
    return this.get(new HttpParams().set("serviceTypeIds", serviceTypes.map(st => st.id).join(',')).set("serviceId", service.id), "service/modify");
  }

}

File: src/main/java/com/jss/osiris/modules/quotation/service/ServiceService.java
package com.jss.osiris.modules.quotation.service;

import java.util.List;

import com.jss.osiris.libs.exception.OsirisException;
import com.jss.osiris.modules.quotation.model.Affaire;
import com.jss.osiris.modules.quotation.model.Service;
import com.jss.osiris.modules.quotation.model.ServiceType;

public interface ServiceService {
    public Service getService(Integer id);

    public Service addOrUpdateService(Service service);

    public Service modifyServiceType(ServiceType serviceType, Service service);

    public String getServiceLabel(Service service) throws OsirisException;

    public Service getServiceForMultiServiceTypesAndAffaire(List<ServiceType> serviceTypes, Affaire affaire);
}
